File: Money/Money.js
export default class Money {
	constructor(value, currency) {
		this._amount = value;
		this._currency = currency;
	}
	
	equals(object) {
		return this._amount == object._amount && this.getCurrency() == object.getCurrency();
	}
	
	times(multiplier) {
		return new Money(this._amount * multiplier, this.getCurrency());
	}
	
	getCurrency() {
		return this._currency;
	}
	
	static dollar(value) {
		return new Money(value, 'USD');
	}
	
	static won(value) {
		return new Money(value, 'KRW');
	}
}

File: Java/Money/app/src/main/java/Money/Money.java
package Money;

class Money implements Expression {
    protected int amount;
    protected String currency;

    Money(int amount, String currency) {
        this.amount = amount;
        this.currency = currency;
    }

    static Money dollar(int amount) {
        return new Money(amount, "USD");
    }

    static Money won(int amount) {
        return new Money(amount, "KRW");
    }

    Money times(int multiplier) {
        return new Money(amount * multiplier, currency);
    }

    Expression plus(Money addend) {
        return new Sum(this, addend);
    }

    public boolean equals(Object object) {
        Money money = (Money) object;
        return amount == money.amount && currency().equals(money.currency());
    }

    public Money reduce(Bank bank, String to) {
        int rate = bank.rate(currency, to);
        return new Money(amount / rate, to);
    }

    String currency() {
        return currency;
    };

    public String toString() {
        return amount + " " + currency;
    }
}
